DIPs/DIP1xxx-rval_to_ref.md

```diff
@@ -0,0 +1,345 @@
+```
+Necessitating the workaround:
+```d
+int temp = 10;
+fun(temp);
+```
+In practise, the argument would likely be some larger struct type rather than 'int', but the inconvenience applies generally.
+
+This inconvenience also extends more broadly to cases including:
+```d
+fun(10);       // literals
+fun(gun());    // return values from functions
+fun(x.prop);   // properties
+fun(x + y);    // expressions
+fun(my_short); // implicit type conversions (ie, short->int promotion)
+// etc... (basically, most things you pass to functions)
+```
+The work-around can bloat the number of lines around the call-site significantly, and the user needs to declare names for all the temporaries, polluting the local namespace, and often for expressions where no meaningful name exists, leading to.
+
+The generic case may appear in a form like this:
+```d
+void someMeta(alias userFun)()
+{
+    userFun(getValue());
+}
+
+void fun(int x);
+void gun(ref const(int) x);
+
+unittest
+{
+    someMeta!(fun)(); // no problem
+    someMeta!(gun)(); // oh no, can't receive rvalue!
+}
+```
+Necessitating a workaround that may look like:
+```d
+void someMeta(alias userFun)()
+{
+    std.algorithm : canFind;
+    static if(canFind(__traits(getParameterStorageClasses, userFun, 0), "ref"))
+    {
+        auto x = getValue();
+        userFun(x);
+    }
+    else
+    {
+        userFun(getValue());
+    }
+}
+```
+This example situation is simplified, but it is often that such issues appear in complex aggregate meta, which may be difficult to understand, or the issue is caused indirectly at some layer the user did not author.
+
+These work-arounds damage readability and brevity, they make authoring correct code more difficult, increase the probability of brittle meta, and it's frustrating to implement repeatedly.
+
+## Proposal
+
+Calls with `ref T` arguments supplied with rvalues are effectively rewritten to emit a temporary automatically, for example:
+```d
+void fun(ref int x);
+
+fun(10);
+```
+Is rewritten:
+```d
+{
+  T __temp0 = void;
+  fun(__temp0 := 10);
+}
+```
+Where `T` is the *function argument type*.
+
+To mitigate confusion, I have used `:=` in this example to express the initial construction, and not a copy operation as would be expected if this code were written with an `=` expression.
+
+In the case where a function output initialises an variable:
+```d
+R result = fun(10);
+```
+Becomes:
+```d
+R result = void;
+{
+  T __temp0 = void;
+  result := fun(__temp0 := 10);
+}
+```
+Again, where initial construction of `result` should be performed at the moment of assignment, as usual and expected.
+
+It is important that `T` be defined as the argument type, and not `auto`, because it will allow implicit conversions to occur naturally, with identical behaviour as when argument was not `ref`. The user should not experience edge cases, or differences in functionality when calling `fun(int x)` vs `fun(ref int x)`.
+
+## Temporary destruction
```

Comment:
This should be a level 3 subheading under description, as should all the headings between here and the copyright.